### PR TITLE
8293293: Move archive heap loading code out of heapShared.cpp

### PR DESCRIPTION
**Pull request by iklam**
I moved all code related to loading the archive heap regions into a new file, archiveHeapLoader.cpp.

A diff of the new archiveHeapLoader.cpp and the old heapShared.cpp shows that the moved code is identical, except for the change of `HeapShared::` to `ArchiveHeapLoader::`.

I also removed unnecessary entries in JVM_EXCLUDE_FILES that are already covered by the `cds/` pattern.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8293293](https://bugs.openjdk.org/browse/JDK-8293293): Move archive heap loading code out of heapShared.cpp

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [35c14e8d](https://git.openjdk.org/jdk/pull/10138/files/35c14e8d6969eddf6cf7c8684bc3769e712de145)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10138/head:pull/10138` \
`$ git checkout pull/10138`

Update a local copy of the PR: \
`$ git checkout pull/10138` \
`$ git pull https://git.openjdk.org/jdk pull/10138/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10138`

View PR using the GUI difftool: \
`$ git pr show -t 10138`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10138.diff">https://git.openjdk.org/jdk/pull/10138.diff</a>

</details>

### CODE REVIEW

**Review by erikj79 — APPROVED** [review]
Build change looks good.

File: src/hotspot/share/cds/archiveHeapLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+#include "memory/metadataFactory.hpp"
+#include "memory/metaspaceClosure.hpp"
+#include "memory/resourceArea.hpp"
+#include "memory/universe.hpp"
+#include "oops/compressedOops.inline.hpp"
+#include "oops/fieldStreams.inline.hpp"
+#include "oops/objArrayOop.inline.hpp"
+#include "oops/oop.inline.hpp"
+#include "oops/typeArrayOop.inline.hpp"
+#include "prims/jvmtiExport.hpp"
```
Comment (coleenp) [inline src/hotspot/share/cds/archiveHeapLoader.cpp:58]: It looks like you should be able to remove a lot of these header files.

**Review by coleenp — APPROVED** [review]
Great! looks good

File: src/hotspot/share/cds/archiveHeapLoader.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/archiveHeapLoader.cpp]: Thanks for the review. I took out the unnecessary includes in the four affected files {heapShared,archiveHeapLoader}.{cpp,hpp}. As a result, I had to fix a few other files to add missing header files that were included transitively.

**Comment by iklam** [comment]
Thanks for the review @erikj79 @coleenp 
/integrate

**Pull request closed without merging**